librz/util/subprocess.c

```diff
@@ -0,0 +1,791 @@
+		if (!stdout_eof) {
+			stdout_index = handles.len;
+			rz_vector_push (&handles, &stdout_overlapped.hEvent);
+		}
+		if (!stderr_eof) {
+			stderr_index = handles.len;
+			rz_vector_push (&handles, &stderr_overlapped.hEvent);
+		}
+		if (!child_dead) {
+			proc_index = handles.len;
+			rz_vector_push (&handles, &proc->proc);
+		}
+
+		DWORD timeout = INFINITE;
+		if (timeout_us_abs != UT64_MAX) {
+			ut64 now = rz_time_now_mono ();
+			if (now >= timeout_us_abs) {
+				return false;
+			}
+			timeout = (DWORD)((timeout_us_abs - now) / RZ_USEC_PER_MSEC);
+		}
+		DWORD signaled = WaitForMultipleObjects (handles.len, handles.a, FALSE, timeout);
+		if (!stdout_eof && signaled == stdout_index) {
+			DWORD r;
+			BOOL res = GetOverlappedResult (proc->stdout_read, &stdout_overlapped, &r, TRUE);
+			if (!res) {
+				stdout_eof = true;
+				continue;
+			}
+			stdout_buf[r] = '\0';
+			rz_str_remove_char (stdout_buf, '\r');
+			rz_strbuf_append (&proc->out, (const char *)stdout_buf);
+			ResetEvent (stdout_overlapped.hEvent);
+			DO_READ (stdout)
+			continue;
+		}
+		if (!stderr_eof && signaled == stderr_index) {
+			DWORD read;
+			BOOL res = GetOverlappedResult (proc->stderr_read, &stderr_overlapped, &read, TRUE);
+			if (!res) {
+				stderr_eof = true;
+				continue;
+			}
+			stderr_buf[read] = '\0';
+			rz_str_remove_char (stderr_buf, '\r');
+			rz_strbuf_append (&proc->err, (const char *)stderr_buf);
+			ResetEvent (stderr_overlapped.hEvent);
+			DO_READ (stderr);
+			continue;
+		}
+		if (!child_dead && signaled == proc_index) {
+			child_dead = true;
+			DWORD exit_code;
+			if (GetExitCodeProcess (proc->proc, &exit_code)) {
+				proc->ret = exit_code;
+			}
+			continue;
+		}
+		break;
+	}
+	rz_vector_clear (&handles);
+	CloseHandle (stdout_overlapped.hEvent);
+	CloseHandle (stderr_overlapped.hEvent);
+	return stdout_eof && stderr_eof && child_dead;
+}
+
+RZ_API void rz_subprocess_kill(RzSubprocess *proc) {
+	TerminateProcess (proc->proc, 255);
+}
+
+RZ_API void rz_subprocess_stdin_write(RzSubprocess *proc, const ut8 *buf, size_t buf_size) {
+	DWORD read;
+	WriteFile (proc->stdin_write, buf, buf_size, &read, NULL);
+}
+
+
+RZ_API RzSubprocessOutput *rz_subprocess_drain(RzSubprocess *proc) {
+	RzSubprocessOutput *out = RZ_NEW (RzSubprocessOutput);
+	if (!out) {
+		return NULL;
+	}
+	out->out = rz_strbuf_drain_nofree (&proc->out);
+	out->err = rz_strbuf_drain_nofree (&proc->err);
+	out->ret = proc->ret;
+	return out;
+}
+
+RZ_API void rz_subprocess_free(RzSubprocess *proc) {
+	if (!proc) {
+		return;
+	}
+	CloseHandle (proc->stdin_write);
+	CloseHandle (proc->stdout_read);
+	CloseHandle (proc->stderr_read);
+	CloseHandle (proc->proc);
+	free (proc);
+}
+
+RZ_API int rz_subprocess_ret(RzSubprocess *proc) {
+	return proc->ret;
+}
+
+RZ_API char *rz_subprocess_out(RzSubprocess *proc) {
+	return rz_strbuf_drain_nofree (&proc->out);
+}
+
+RZ_API char *rz_subprocess_err(RzSubprocess *proc) {
+	return rz_strbuf_drain_nofree (&proc->err);
+}
+#else
```

Comment:
Please add here a comment of what `#if[def]` branch it's end since the blocks are big.